Doc chapter on how to lex string literals (#654)
lalrpop/lalrpop: doc/src/lexer_tutorial/002_raw_delimited_content.md

# Lexing raw delimited content

Our calculator example operated on numbers and arithmetic operators.
There is no overlap between the characters for numeric digits (`0`, `1`, ...)
the characters representing operators (`+`, `-`, ...) and parentheses
(`(`, `)`), so it was easy to embed those tokens directly in the grammar,
as we saw in the earlier sections.

However, some languages do not have such easy separations at the lexical level.

> Consider parsing a language with string literals. We will deliberate choose

Author: ```suggestion
Consider parsing a language with string literals. We will deliberately choose
```

    assert_eq!(nobol5::StmtParser::new().parse(r#"y = "bc""#), Ok(('y', "bc").into()));
    assert_eq!(nobol5::StmtParser::new().parse(r#"z = "xyz""#), Ok(('z', "xyz").into()));
    assert_eq!(nobol5::StmtParser::new().parse(r#"z = "x""#), Ok(('z', "x").into()));
    assert_eq!(nobol5::StmtParser::new().parse(r#"z = "x y z""#), Ok(('z', "x y z").into()));
}
```

Furthermore, we can now remove other artificial limits in our language.
For example, we can make our identifiers more than one character:

```lalrpop
pub Var: Var = <r"[a-z]+"> => <>.into()
```

which, with suitable changes to the library code, works out fine.

#### Escape sequences

Our current string literals are allowed to hold a small subset of the full space of characters.

If we wanted to generalize it to be able to hold arbitrary characters, we would
need some way to denote the delimiter character `"` in ths string content.

The usual way to do this is via an escape sequence: `\"`, which is understood by
the lexical analyzer as *not* ending the string content.

We can generalize the regular expression in our new `Lit` rule to handle this:

```lalrpop

> pub Lit: Lit = <l:r#""(\\\\|\\"|[^"])*""#> => l[1..l.len()-1].into();

Author: ```suggestion
pub Lit: Lit = <l:r#""(\\\\|\\"|[^"\\])*""#> => l[1..l.len()-1].into();
```

```

However, depending on your data model, this is not quite right. In particular:
the string is produced still has the escaping backslashes embedded in it.

As a concrete example, with the above definition for `Lit`, this test:

```rust
#[test]
fn popfly() {
    assert_eq!(nobol6::StmtParser::new().parse(r#"z = "\"\\""#), Ok(('z', "\"\\").into()));
}
```

yields this output:

```
thread 'popfly' panicked at 'assertion failed: `(left == right)`
  left: `Ok(Stmt(Var('z'), Lit("\\\"\\\\")))`,
 right: `Ok(Stmt(Var('z'), Lit("\"\\")))`', doc/nobol/src/main.rs:91:5
```

This can be readily addressed by adding some code to post-process the token to remove the
backslashes:

```lalrpop

> pub Lit: Lit = <l:r#""(\\\\|\\"|[^"])*""#> =>
>     l[1..l.len()-1].replace("\\\"", "\"").replace("\\\\", "\\").into();

Author: ```suggestion
pub Lit: Lit = <l:r#""(\\\\|\\"|[^"\\])*""#> => Lit(apply_string_escapes(&l[1..l.len()-1]).into());
```

> ```

Author: ```suggestion
```

where `apply_string_escape` is a helper routine that searches for occurrences of `\\` and does the appropriate substitution within the string content.
```

> ```

Author: ````suggestion
```

where `apply_string_escapes` is a helper routine that searches for backslashes in the content and performs the corresponding replacement with the character denoted by the escape sequence.
````

# Lexing raw delimited content

Our calculator example operated on numbers and arithmetic operators.

> There is no overlap between the characters for numeric digits (`0`, `1`, ...)

Reviewer: ```suggestion
There is no overlap between the characters for numeric digits (`0`, `1`, ...),
```

the characters representing operators (`+`, `-`, ...) and parentheses
(`(`, `)`), so it was easy to embed those tokens directly in the grammar,
as we saw in the earlier sections.

> However, some languages do not have such easy separations at the lexical level.

Reviewer: ```suggestion
However, some languages do have overlapping lexical token.
```

Author: Addressed in fb568a3be286438944569aae7d25b33e6e0fa17b

Consider parsing a language with string literals. We will deliberately choose
one that is very simple: All it can do is bind variables, and the variables are

> always single characters, like this:

Author: Addressed in fb568a3be286438944569aae7d25b33e6e0fa17b

```
x = "a"
y = "bc"
```

> Using what we have learned so far, we might try a grammar like the below.

Reviewer: ```suggestion
Using what we have learned so far, we might try a grammar like the following one:
```

fn nobol1() {
    assert_eq!(nobol1::VarParser::new().parse("x"), Ok('x'.into()));
    assert_eq!(nobol1::LitParser::new().parse(r#""abc""#), Ok("abc".into()));
    assert_eq!(nobol1::StmtParser::new().parse(r#"x = "a""#), Ok(('x', "a").into()));
    assert_eq!(nobol1::StmtParser::new().parse(r#"y = "bc""#), Ok(('y', "bc").into()));
}
```

However, as soon as we try to handle examples where there *is* overlap between
the variable names and the string contents, we run into a problem.

Consider an example like this:

```
z = "xyz"
```

If we encode that as a test,

```rust
#[test]
fn no_ball() {
    assert_eq!(nobol1::StmtParser::new().parse(r#"z = "xyz""#), Ok(('z', "xyz").into()));
}
```

we get the following:

```
thread 'no_ball' panicked at 'assertion failed: `(left == right)`
  left: `Err(UnrecognizedToken { token: (5, Token(1, "x"), 6), expected: ["r#\"[a-c]*\"#"] })`,
 right: `Ok(Stmt(Var('z'), Lit("xyz")))`', doc/nobol/src/main.rs:31:5
```

Okay, this sounds easy to fix: the regular expression for string literal content should
be expanded to handle characters other than `a`, `b` and `c`.

So does this work?

```lalrpop
pub Lit: Lit = "\"" <r"[a-z]*"> "\"" => <>.into();

> ```

Reviewer: I think this section can be cut out. The guide starts by saying that the calculator had a clear separation between lexical tokens, but this document is about grammars with some form of overlapping. However, we start the section with a grammar without overlapping, and which honestly looks really artificial. I'm not sure it adds a lot of value.

My suggestion would be to get rid of this first step and directly starts with imagining a language with strings, variables and statements, and then write the naive ambiguous version of this grammar.

Author: Its been over a year, but from what I remember about the way that my brain works, it was easiest for *me* to get the initial grammar working by starting from the artificially limited case, get all those pieces to work, then figure out how to generalize it.

Sure, the way I restricted it was absurdly limited. But there is plenty of precedent, especially in PL work, of people forcing programmers to follow seemingly absurd syntactic restrictions in order to satisfy the limits of the lexer/parser implementation.

Anyway, my point is more that I thought it was worthwhile to spell out "here's working code, but it doesn't readily generalize." And it *is* worthwhile having those artificial limits when it makes the discussion clear; for example, in the section "The right way to do this", the places where it refers to `r"[a-c]*"` is entirely *unambiguous* because that range is solely used in the definition for the `Lit` production.

Author: Having said all that, let me take a shot at the re-write you suggest. Maybe it will come out cleaner.

Author: Addressed in fb568a3be286438944569aae7d25b33e6e0fa17b

}
```

The above produces:

```
---- foul_ball stdout ----
thread 'foul_ball' panicked at 'assertion failed: `(left == right)`
  left: `Err(UnrecognizedToken { token: (5, Token(3, "x"), 6), expected: ["r#\"[a-z]*\"#"] })`,
 right: `Ok(Stmt(Var('z'), Lit("x")))`', doc/nobol/src/main.rs:43:5
```

What is the problem?

Merely specifying a precedence to favor tokenizing `r"[x-z]"` over `r"[a-z]*"`
does not address the real problem here. That precedence rule causes an input
like `z = "x"` to be split into tokens such that the `x` only matches the
regular expression for the `Var`. It will not match the `r"[a-z]*"` in the `Lit`
rule, even if it intuitively seems like it should; they have already been
lexically categorized as different tokens at this point.

One could add further workarounds to deal with this. For example, one could
change the `Lit` production to explicitly handle the `r"[x-z]"` regular
expression as its own case:

```lalrpop
pub Lit: Lit = {
    "\"" <r"[x-z]"> "\"" => <>.into(),
    "\"" <r"[a-z]*"> "\"" => <>.into(),
};
```

But this is a fragile workaround.

Specifically, this workaround is only applicable because we put artificial

> limits in this language.

Reviewer: ```suggestion
limits on this language.
```

Not sure though, I'm not a native speaker :slightly_smiling_face:

If we wanted to generalize string literals to be able to contain other
characters (such as whitespace), the technique described so far does not work
out well. Consider this grammar:

```lalrpop
match {
   r"[x-z]"
} else {
   r"[a-z ]*",
   _
}

pub Var: Var = <r"[x-z]"> => <>.chars().next().unwrap().into();

pub Lit: Lit = {
    "\"" <r"[x-z]"> "\"" => <>.into(),
    "\"" <r"[a-z ]*"> "\"" => <>.into(),
};

pub Stmt: Stmt = <Var> "=" <Lit> => (<>).into();
```

Now, if we run the same test as before:

```rust
#[test]
fn spaceballs() {
    assert_eq!(nobol4::StmtParser::new().parse(r#"z = "x""#), Ok(('z', "x").into()));
}
```

we get the following error output:

```
thread 'spaceballs' panicked at 'assertion failed: `(left == right)`
  left: `Err(UnrecognizedToken { token: (0, Token(2, "z "), 2), expected: ["r#\"[x-z]*\"#"] })`,
 right: `Ok(Stmt(Var('z'), Lit("x")))`', doc/nobol/src/main.rs:58:5
```

> Our attempt to generalize what string contents can be has caused problems for

Reviewer: ```suggestion
Our attempt to generalize what strings can contain has caused problems for
```

how the *rest* of the input is tokenized.

#### The right way to do this

Let us revisit the original rule in the grammar for string literals, from our
first version:

```lalrpop
pub Lit: Lit = "\"" <r"[a-c]*"> "\"" => <>.into();
```

The heart of our problem is that we have implicitly specified distinct tokens
for the string delimiter (`"\""`) versus the string content (in this case,
`r"[a-c]*"`).

> The problem is that our intuition is that we only want to tokenize string

Reviewer: ```suggestion
Our intuition is that we only want to tokenize string
```

Author: Addressed in fb568a3be286438944569aae7d25b33e6e0fa17b

content when we are in the process of reading a string. In other words, we only
want to apply the `r"[a-c]*"` rule immediately after reading a `"\""`. But the

> generated lexer does not infer this from our rules; it just blindly pattern
> matches for something matching the string content regular expression *anywhere*
> in the input.

Reviewer: ```suggestion
generated lexer does not infer this from our rules; it just blindly looks for
something matching the string content regular expression *anywhere*
in the input.
```

You could resolve this with a custom lexer (treated in the next section).

> But the simplest thing to do in this scenario is to read the string delimiters and the string content as a *single token*, like so:

Reviewer: ```suggestion
But the simplest thing to do is to read the string delimiters and the string content as a *single token*, like so:
```

Author: Addressed in fb568a3be286438944569aae7d25b33e6e0fa17b
